app/api-client.js

```diff
@@ -1,170 +1,291 @@
+    this.token = token || 'invalid token';
+  }
+
+  set_token(token) {
+    this.token = token;
+  }
+
+  unset_token() {
+    this.token = 'invalid token';
+  }
+
+  get header() {
+    return `JWT ${this.token}`;
+  }
+}
+
+
+class APIClient {
+  constructor(conf) {
+    this.base_url = conf.base_url;
+    this.auth = null;
+  }
+
+  request(endpoint, method = 'GET', body = null) {
+    const headers = {};
+
+    if (this.auth) {
+      headers.Authorization = this.auth.header;
+    }
 
-  function api_request(options) {
     try {
-      return request(override_defaults(options));
+      const options = {
+        method,
+        uri: this.endpoint_url(endpoint),
+        headers,
+        body,
+        json: true,
+      };
+      return request(options)
+        .then((result) => {
+          log.debug(`${method} ${options.uri}`);
+          // console.dir(result);
+          return result;
+        });
     } catch (error) {
-      throw new Error(`API: ${options.method || 'GET'} ${options.endpoint} failed: ${error}`);
+      throw new Error(`API: ${method} ${endpoint} failed: ${error}`);
     }
   }
 
-  function override_defaults(options) {
-    let defaults = {
-      method: 'GET',
-      uri: endpoint_url(options.endpoint),
-      headers: {
-        Authorization: jwt_auth(),
-      },
-      body: options.resource,
-      json: true,
-    };
+  get(endpoint) {
+    return this.request(endpoint);
+  }
+
+  post(endpoint, body = '') {
+    return this.request(endpoint, 'POST', body);
+  }
 
-    return Object.assign(defaults, options);
+  delete(endpoint) {
+    return this.request(endpoint, 'DELETE');
   }
 
-  function endpoint_url(endpoint) {
+  patch(endpoint, body = '') {
+    return this.request(endpoint, 'PATCH', body);
+  }
+
+  put(endpoint, body = '') {
+    return this.request(endpoint, 'PUT', body);
+  }
+
+  endpoint_url(endpoint) {
     if (!endpoint) {
       throw new Error('Missing endpoint');
     }
 
+    let suffix = '';
     if (!endpoint.endsWith('/')) {
-      endpoint += '/';
+      suffix = '/';
     }
 
-    return url.resolve(config.api.base_url, endpoint);
+    return url.resolve(this.base_url, endpoint + suffix);
   }
+}
+
+
+const api = new APIClient(config.api);
+api.auth = new JWTAuth();
+
+exports.api = api;
+
 
-  function basic_auth() {
-    return 'Basic ' + new Buffer(
-      config.api.username + ':' + config.api.password).toString('base64');
+const model_proxy = {
+  get: (model, property) => {
+    if (Reflect.has(model, property)) {
+      return Reflect.get(model, property);
+    }
+    if (model.data && property in model.data) {
+      return model.data[property];
+    }
+    return undefined;
+  },
+};
+
+
+class ModelSet extends Array {
+  constructor(Model, data = []) {
+    super(...data.map(obj => new Model(obj)));
+    this.model = Model;
   }
 
-  function jwt_auth() {
-    return `JWT ${token}`;
+  exclude(other) {
+    let others = other;
+    if (other instanceof this.model) {
```

Comment:
You may be able to get rid of the `if` by using the `Array` constructor with the spread, with something like `const others = Array(...other)` (untested but should do).

Reply:
Nope, that would fail if `other` is not an iterable.

Reply:
Sorry, I tried this here and it worked but maybe I'm missing the context here: https://jsfiddle.net/#&togetherjs=voaHu070qF